### ISSUE TITLE
DECLARE_BITMAP with array initialization breaks spatch

### ISSUE THREAD

**Issue by jacob-keller** [coccinelle/coccinelle#167]
I have some code which uses DECLARE_BITMAP(name, size) = { 0 };

This code is not parsed correctly, because DECLARE_BITMAP doesn't have a proper definition. which results in the functions which have this type of invocation not being parsed correctly.

This seems to be the result of commit 511277cd7ac7 ("not clear why DECLARE_BITMAP needs a definition when the others dont have one", 2010-12-18)

I think most of the other DECLARE_* style macros aren't as likely to be used with static initializers like this.

Should we just use our own macro file copy, or is this something worth fixing upstream?

**Comment by jacob-keller** [comment]
it looks like to fix this we need to pass --macro-file-builtin with a modified copy of the standard.h

There doesn't appear to be a way to override only one standard.h macro.. that would be most convenient for us.

**Comment by jacob-keller** [comment]
We can use just a separate macro-file since commit a02ce63bed6e ("drop some declares that are no longer useful", 2018-05-06)

**Comment by JuliaLawall** [comment]
On Thu, 30 May 2019, Jacob Keller wrote:

>
> I have some code which uses DECLARE_BITMAP(name, size) = { 0 };
>
> This code is not parsed correctly, because DECLARE_BITMAP doesn't have a
> proper definition. which results in the functions which have this type of
> invocation not being parsed correctly.
>
> This seems to be the result of commit 511277c ("not clear why DECLARE_BITMAP
> needs a definition when the others dont have one", 2010-12-18)
>
> I think most of the other DECLARE_* style macros aren't as likely to be used
> with static initializers like this.
>
> Should we just use our own macro file copy, or is this something worth
> fixing upstream?

I just uncommented the definition.  Thanks for the report.

julia

**Comment by JuliaLawall** [comment]
On Thu, 30 May 2019, Jacob Keller wrote:

>
> it looks like to fix this we need to pass --macro-file-builtin with a
> modified copy of the standard.h
>
> There doesn't appear to be a way to override only one standard.h macro..
> that would be most convenient for us.

I think that --macro-file is a file of definitions that extend standard.h,
but I'm not sure.  In any case, perhaps you can just install the github
version?

julia

**Comment by jacob-keller** [comment]
> On Thu, 30 May 2019, Jacob Keller wrote: it looks like to fix this we need to pass --macro-file-builtin with a modified copy of the standard.h There doesn't appear to be a way to override only one standard.h macro.. that would be most convenient for us.
> I think that --macro-file is a file of definitions that extend standard.h, but I'm not sure.

Well, I tried using --macro-file with a header that only defined DECLARE_BITMAP. This worked for the newest 1.0.7 version (prior to your patch).

However, for older versions the standard.h definition of DECLARE_BITMAP got used. I tried to search the code to understand how or why that happened, but it wasn't obvious to me.

It would be nice if --macro-file definitions overrode definitions in standard.h, since this would mean not having to rely on keeping our copy of standard.h up to date if we run into a similar issue with other macros in the future.

Thanks for the help!
